### PR TITLE
Extend hardhat-etherscan with ability to verify proxy

### PR DESCRIPTION
**Pull request by ericglau**
Overrides hardhat-etherscan's `verify` command so that when a proxy is detected, attempts full verification of all proxy-related contracts including implementation, proxy admin (if any), beacon (if any), and proxy (UUPS or transparent or beacon proxy).  Also links the proxy to implementation ABI using Etherscan's verify as proxy API.

Fixes https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/242

### CODE REVIEW

**Comment by alcuadrado** [comment]
Hey @ericglau,

I found this just by chance. Let us know if we can help in any way.

In our experience, the main pain point that users have when verifying OZ upgradable contracts is that they always try to verify the proxy and not the implementation.

**Comment by ericglau** [comment]
Thanks @alcuadrado.  This PR will override hardhat-etherscan's `verify` command to detect the implementation address from a proxy and run `verify` on the actual implementation address directly (which should address the pain point that you mention), in addition to verifying the actual proxy contracts for the user.

**Comment by alcuadrado** [comment]
That's awesome!

**Review by frangio — COMMENTED** [review]
Great work!

File: packages/plugin-hardhat/src/verify-proxy.ts [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+  if (await isTransparentOrUUPSProxy(provider, proxyAddress)) {
+    await fullVerifyTransparentOrUUPS(hre, proxyAddress, hardhatVerify);
+  } else if (await isBeaconProxy(provider, proxyAddress)) {
+    await fullVerifyBeaconProxy(hre, proxyAddress, hardhatVerify);
+  } else {
+    // Doesn't look like a proxy, so just verify directly
+    return hardhatVerify(proxyAddress);
+  }
+
+  if (errors.length > 0) {
```
[the remark this answers is not in the document]
Comment (ericglau) [reply packages/plugin-hardhat/src/verify-proxy.ts:115]: Changed it to pass the errors array as function arguments.

File: packages/plugin-hardhat/src/verify-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+ */
+interface ContractEventMapping {
+  artifact: ContractArtifact;
+  event: string;
+}
+
+/**
+ * Types of proxy-related contracts and their corresponding events.
+ */
+interface ContractEventsMap {
```
Comment (frangio) [inline packages/plugin-hardhat/src/verify-proxy.ts:58]: I find the type names confusing. I would rename:

- `ContractEventsMap` -> `VerifiableContracts`
- `ContractEventsMapping` -> `VerifiableContractInfo`

**Review by frangio — APPROVED** [review]
LGTM!

**Pull request merged**